src/v/cluster/metadata_cache.h

```diff
@@ -118,6 +118,8 @@ class metadata_cache {
     /// broker can change
     std::optional<broker_ptr> get_broker(model::node_id) const;
 
+    bool should_reject_writes() const;
```

Comment:
Ah, I've been thinking of metadata cache as just that, for the Kafka layer (which is the caller). I'm getting the impression that metadata cache is more like a specific facade, not for _metadata needed by the kafka layer_ (the natural _code_ structure in my mind), but for _metadata needed to answer Kafka Metadata API queries_ only. Is that more accurate?

Ultimately I'll defer to you guys here but I think some background might help...

My bias from past experience is that we need to unify our cluster quorum metadata API into one consistent/coherent place. Metadata cache API is probably the closest to that so far, from perspective of kafka layer, but it is not coherent[1]. IMO disk health needs to be coherent with node health/status or you need to implement retries and really make your code more complex--so I have a hard time understanding why you'd want them separated. A whiteboard and some war stories might help illustrate my point better. Maybe I'm being too strict? Basically, a "healthy node in your cluster" is not really healthy if it doesn't have any storage space. And to scale your cluster level-quorum (e.g. controller) past 100 nodes, you need to restrict it to the set of stuff you need to know, coherently, to start the request processing path. Things like nodes and disks live here because you can't access other distributed structures (logs, indexes, btrees, etc) without a pointer to them. And a (node,disk,addr) are required for that "pointer". Does that make any sense? 😅 These simplifying assumptions (i.e. I can get coherent node+disk metadata that will be automatically valid for the life of my request--with automatic recovery retries on quorum change) might not be 100% required, but they can make our lives much easier, and be rock solid under abusive chaos/jepsen-style testing. 

[1]  For example, we have stuff [like this](https://github.com/redpanda-data/redpanda/blob/647eeb68c9a52de6ffb0c6b0be980afa4c55a1ab/src/v/cluster/metadata_cache.cc#L119-L126), which is incorrect (IIUC).

Reply:
ah i wasn't aware of that linked source there. But my opinion is that this `all_alive_brokers` method shouldn't be in the metadata cache either.

Reply:
How about a followup PR on this one? We can get consensus on next steps for the cluster metadata & health stuff and go from there. (Either plumb health monitor into kafka layer, or settle on a new coherent API and build that. I have some ideas. 😉)

Reply:
Sounds good to me